swarm/network/stream/snapshot_sync_test.go

```diff
@@ -144,29 +95,185 @@ func TestSyncing(t *testing.T) {
+			}
+		}
+	}
+}
+
 func testSyncing(t *testing.T, chunkCount int, nodeCount int) {
-	//test live and NO history
-	log.Info("Testing live and no history")
-	live = true
-	history = false
-	err := runSyncTest(chunkCount, nodeCount, live, history)
+	bucketKeyStore := simulation.BucketKey("store")
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"streamer": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+
+			id := ctx.Config.ID
+			addr := network.NewAddrFromNodeID(id)
+			store, datadir, err := createTestLocalStorageForId(id, addr)
+			if err != nil {
+				return nil, nil, err
+			}
+			bucket.Store(bucketKeyStore, store)
+			cleanup = func() {
+				store.Close()
+				os.RemoveAll(datadir)
+			}
+			localStore := store.(*storage.LocalStore)
+			db := storage.NewDBAPI(localStore)
+			kad := network.NewKademlia(addr.Over(), network.NewKadParams())
+			delivery := NewDelivery(kad, db)
+			bucket.Store(bucketKeyDelivery, delivery)
+
+			r := NewRegistry(addr, delivery, db, state.NewInmemoryStore(), &RegistryOptions{
+				SkipCheck:       defaultSkipCheck,
+				DoRetrieve:      false,
+				DoSync:          true,
+				SyncUpdateDelay: 3 * time.Second,
+			})
+			RegisterSwarmSyncerServer(r, db)
+			RegisterSwarmSyncerClient(r, db)
+
+			fileStore := storage.NewFileStore(storage.NewNetStore(localStore, nil), storage.NewFileStoreParams())
+			testRegistry := &TestRegistry{Registry: r, fileStore: fileStore}
+			bucket.Store(bucketKeyRegistry, testRegistry)
+
+			return testRegistry, cleanup, nil
+
+		},
+	})
+	defer sim.Close()
+
+	log.Info("Initializing test config")
+
+	ctx := context.Background()
+
+	conf := &synctestConfig{}
+	//map of discover ID to indexes of chunks expected at that ID
+	conf.idToChunksMap = make(map[discover.NodeID][]int)
+	//map of overlay address to discover ID
+	conf.addrToIdMap = make(map[string]discover.NodeID)
+	//array where the generated chunk hashes will be stored
+	conf.hashes = make([]storage.Address, 0)
+	//time.Sleep(5 * time.Second)
+
+	err := sim.UploadSnapshot(fmt.Sprintf("testing/snapshot_%d.json", nodeCount))
 	if err != nil {
 		t.Fatal(err)
 	}
-	//test history only
-	log.Info("Testing history only")
-	live = false
-	history = true
-	err = runSyncTest(chunkCount, nodeCount, live, history)
-	if err != nil {
-		t.Fatal(err)
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		nodeIDs := sim.UpNodeIDs()
+		for _, n := range nodeIDs {
+			//get the kademlia overlay address from this ID
+			a := network.ToOverlayAddr(n.Bytes())
+			//append it to the array of all overlay addresses
+			conf.addrs = append(conf.addrs, a)
+			//the proximity calculation is on overlay addr,
+			//the p2p/simulations check func triggers on discover.NodeID,
+			//so we need to know which overlay addr maps to which nodeID
+			conf.addrToIdMap[string(a)] = n
+		}
+
+		//select one index at random...
+		idx := rand.Intn(len(nodeIDs))
```

Comment:
This looks to me that we need sim.RandomNode() function. There already is, but not exposed. Do you think that it should be exposed to simplify this part of the test?

Reply:
well it's two lines of code but why not :) it would be helpful

Reply:
Great, I will create a small PR.

Reply:
Submitted https://github.com/holisticode/go-ethereum/pull/2.